File: k_small_bst.py
# Time complexity - O(k+h) where h:height of the tree 
# Space Complexity: O(h) 
# Did this code successfully run on LeetCode?: Yes
# Problems faced while coding this:None
# Approach : Iterative method - traverse to the left side as we are searching for kth smallest in the BST and decrement k to reach the appropriate node.

class Solution:
    def kthSmallest(self, root: TreeNode, k: int) -> int:
        
        if root is None:
            return -1
        
        stack =[]
        while root is not None or len(stack)>0:
            
            while root is not None:
                stack.append(root)
                root=root.left
            
            root = stack.pop()
            k-=1
            if k==0:
                return root.val
            root = root.right
        
        return -1
        
        
        

File: ksmall_bst.cpp
//TC: O(n) where n is the number of nodes in the tree
//SC: O(1)

/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int kthSmallest(TreeNode* root, int k) {
        
        TreeNode *curr = root;
        stack<TreeNode*> s;
        
        int ksmall = -1;
        
        //Done using iterative inorder traversal since in inorder traversal, the nodes are in ascending order
        
        while(curr!=NULL || !s.empty()){
            
            while(curr!=NULL){
                s.push(curr);
                curr = curr->left;
            }
            
            curr = s.top();
            s.pop();
            
            ksmall = curr->val;
            k--;
            if(k == 0){
                break;
            }
            
            curr = curr->right;
        }
        
        return ksmall;
        
    }
};